### PR TITLE
CORE-18515: Create gateway HTTP RPC client

### PR DESCRIPTION
**Pull request by yift-r3**
Introduce an HTTP client to be used as RPC client for a `SyncRPCSubscriptionImpl` server without the overhead of the `RPCClient`.

### CODE REVIEW

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/utils/HttpRpcClient.kt [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    private companion object {
+        val logger = LoggerFactory.getLogger(HttpRpcClient::class.java)
+        const val MAX_RETRIES = 3
+        val TIMEOUT = Duration.ofSeconds(30)
+        val INITIAL_DELAY = Duration.ofMillis(100)
+        const val DELAY_INCREASE_FACTOR = 2L
+        private const val SUCCESS = "SUCCESS"
+        private const val FAILED = "FAILED" }
```
Comment (williamvigorr3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/utils/HttpRpcClient.kt]: Nit `}` should be on the next line.

File: components/gateway/src/test/kotlin/net/corda/p2p/gateway/utils/HttpRpcClientTest.kt [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    @Test
+    fun `send post the correct data`() {
+        val body = argumentCaptor<HttpRequest.BodyPublisher>()
+        whenever(requestBuilder.POST(body.capture())).doReturn(requestBuilder)
+
+        client.send<Date>(uri, request)
+
+        assertThat(body.firstValue.contentLength()).isEqualTo(serializedRequest.size.toLong())
```
Comment (williamvigorr3) [inline components/gateway/src/test/kotlin/net/corda/p2p/gateway/utils/HttpRpcClientTest.kt:102]: Nit: Should we compare the content.

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/utils/HttpRpcClient.kt [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                cause = null,
+            )
+    }
+
+    private companion object {
+        val logger = LoggerFactory.getLogger(HttpRpcClient::class.java)
+        const val MAX_RETRIES = 3
+        val TIMEOUT = Duration.ofSeconds(30)
+        val INITIAL_DELAY = Duration.ofMillis(100)
+        const val DELAY_INCREASE_FACTOR = 2L
```
Comment (williamvigorr3) [inline components/gateway/src/main/kotlin/net/corda/p2p/gateway/utils/HttpRpcClient.kt:51]: Where did you get these numbers from? They look Okay to me, I am just curious.

**Review by williamvigorr3 — COMMENTED** [review]
Looks good. I left two minor comments.

File: components/gateway/src/main/kotlin/net/corda/p2p/gateway/utils/HttpRpcClient.kt (hunk 3 above)
Comment (yift-r3) [reply components/gateway/src/main/kotlin/net/corda/p2p/gateway/utils/HttpRpcClient.kt:51]: From `HTTPRetryConfig`

File: components/gateway/src/test/kotlin/net/corda/p2p/gateway/utils/HttpRpcClientTest.kt (hunk 2 above)
Comment (yift-r3) [reply components/gateway/src/test/kotlin/net/corda/p2p/gateway/utils/HttpRpcClientTest.kt:102]: The `BodyPublisher` won't expose the content.

**Review by williamvigorr3 — APPROVED** [review]
LGTM

**Pull request merged**